contractcourt/channel_arbitrator.go

```diff
@@ -1135,11 +1146,141 @@ func (c *ChannelArbitrator) sweepAnchors(anchors *lnwallet.AnchorResolutions,
+			// what's sent from us (toRemoteHTLCs) is the incoming
+			// HTLCs.
+			toLocalHTLCs := htlcs.outgoingHTLCs
+			toRemoteHTLCs := htlcs.incomingHTLCs
+
+			err := sweepWithDeadline(
+				anchors.Remote, toLocalHTLCs, toRemoteHTLCs,
+			)
+			if err != nil {
+				return err
+			}
+
+		case htlcSet == RemotePendingHtlcSet && anchors.RemotePending != nil:
+			// Inside a pending remote commitment transaction,
+			// what's sent to us (toLocalHTLCs) is the outgoing
+			// HTLCs, and what's sent from us (toRemoteHTLCs) is
+			// the incoming HTLCs.
+			toLocalHTLCs := htlcs.outgoingHTLCs
+			toRemoteHTLCs := htlcs.incomingHTLCs
+
+			err := sweepWithDeadline(
+				anchors.RemotePending, toLocalHTLCs,
+				toRemoteHTLCs,
+			)
+			if err != nil {
+				return err
+			}
+		}
 	}
 
 	return nil
 }
 
+// findCommitmentDeadline finds the deadline (relative block height) for a
+// commitment transaction by extracting the minimum CLTV from its HTLCs. From
+// our PoV, the deadline is defined to be the smaller of,
+//   - the least CLTV from toRemoteHTLCs,  or,
+//   - the least CLTV from toLocalHTLCs if the preimage is available.
+// Depending on the commitment transaction type, this means,
+//   - if it's a local commitment transaction,
+//	* toLocalHTLCs is the incoming HTLCs.
+//	* toRemoteHTLCs is the outgoing HTLCs.
+//   - if it's a remote/pending remote commitment transaction,
+//	* toLocalHTLCs is the outgoing HTLCs.
+//	* toRemoteHTLCs is the incoming HTLCs.
+func (c *ChannelArbitrator) findCommitmentDeadline(heightHint uint32,
+	toLocalHTLCs, toRemoteHTLCs map[uint64]channeldb.HTLC) (uint32, error) {
+
+	minHeight := uint32(math.MaxUint32)
+
+	// First, iterate through the toRemoteHTLCs to find the lowest CLTV
+	// value.
+	for _, htlc := range toRemoteHTLCs {
+		// Skip if the HTLC is dust.
+		if htlc.OutputIndex < 0 {
+			log.Debugf("ChannelArbitrator(%v): skipped deadline "+
+				"for dust htlc=%x",
+				c.cfg.ChanPoint, htlc.RHash[:])
+
+			continue
+		}
+
+		if htlc.RefundTimeout < minHeight {
+			minHeight = htlc.RefundTimeout
+		}
+	}
+
+	// Then going through the toLocalHTLCs, and update the minHeight when
+	// conditions met.
+	for _, htlc := range toLocalHTLCs {
+		// Skip if the HTLC is dust.
+		if htlc.OutputIndex < 0 {
+			log.Debugf("ChannelArbitrator(%v): skipped deadline "+
+				"for dust htlc=%x",
+				c.cfg.ChanPoint, htlc.RHash[:])
+
+			continue
+		}
+
+		// Since it's an HTLC sent to us, check if we have preimage for
+		// this HTLC.
+		preimageAvailable, err := c.isPreimageAvailable(htlc.RHash)
+		if err != nil {
+			return 0, err
+		}
+
+		if !preimageAvailable {
+			continue
+		}
+
+		if htlc.RefundTimeout < minHeight {
+			minHeight = htlc.RefundTimeout
+		}
+	}
+
+	// Calculate the deadline. Notice that our minHeight may never get
+	// updated, which means the passed HTLCs set is empty. We will then
+	// fall back to the default value.
+	deadline := minHeight - heightHint
```

Comment:
the height hint here is not necessarily the current block is it?

Reply:
I think it could be one block off if a new block is generated during these lines.

What I see here is,
1. new blocks comes in -> `advanceState` -> `stateStep`, which has the current block, [start from this line](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/channel_arbitrator.go#L2166)
2. the other is from [the loop](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/channel_arbitrator.go#L1174) within `advacenState`, when the state is changed from `StateBroadcastCommit` to `StateCommitmentBroadcasted` from [this line](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/channel_arbitrator.go#L926) after we force close the channel. Depending on how fast these get executed, we might have a new block comes in, and we would be back in case 1.